### PR TITLE
Implement FloatMatrix4x4 class.

### PR DESCRIPTION
**Pull request by IceReaper**
Instead of relying on a hand full of utility methods which manipulate a float[], this pr implements a proper 4x4 float matrix class to use for advanced 3d calculations.

This makes working with 3d content in the future way easier and clearer.

### CODE REVIEW

File: OpenRA.Game/Graphics/ModelRenderer.cs [hunk 1]
```diff
@@ -137,26 +133,26 @@ public void SetViewportParams(Size screen, int2 scroll)
...
+			var shadowScreenTransform = cameraTransform * invShadowTransform;
+			var shadowGroundNormal = shadowTransform * groundNormal;
 			var screenCorners = new float3[4];
 			for (var j = 0; j < 4; j++)
 			{
 				// Project to ground plane
-				corners[j][2] = -(corners[j][1] * shadowGroundNormal[1] / shadowGroundNormal[2] +
-								  corners[j][0] * shadowGroundNormal[0] / shadowGroundNormal[2]);
+				var z = -(corners[j].Y * shadowGroundNormal.Y / shadowGroundNormal.Z +
+				          corners[j].X * shadowGroundNormal.X / shadowGroundNormal.Z);
```
Comment (abcdefg30) [inline OpenRA.Game/Graphics/ModelRenderer.cs]: Spaces.

File: OpenRA.Game/Primitives/FloatMatrix4x4.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
+﻿#region Copyright & License Information
+
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/FloatMatrix4x4.cs]: Superfluous new line here and before `#endregion`.

File: OpenRA.Game/Primitives/FloatMatrix4x4.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		public override int GetHashCode() { return M11.GetHashCode() ^ M22.GetHashCode() ^ M33.GetHashCode() ^ M44.GetHashCode(); }
+
+		public bool Equals(FloatMatrix4x4 other) { return other == this; }
+		public override bool Equals(object obj) { return obj is FloatMatrix4x4 && Equals((FloatMatrix4x4)obj); }
+
+		public override string ToString()
+		{
+			return
+				"[" + M11 + " " + M12 + " " + M13 + " " + M14 + "],[" +
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/FloatMatrix4x4.cs]: This seems to open more `[` than it closes.

Comment (IceReaper) [reply OpenRA.Game/Primitives/FloatMatrix4x4.cs]: Oh! thanks, copied that from Int32Matrix4x4. So both need a fix then :D

**Comment by teinarss** [comment]
IMO we should go for using https://www.nuget.org/packages/System.Numerics.Vectors/

**Pull request closed without merging**